src/jupyterlabcontroller/services/lab.py

```diff
@@ -2,64 +2,201 @@
 
 from __future__ import annotations
 
+import asyncio
+import contextlib
 from base64 import b64encode
-from functools import partial
+from collections.abc import AsyncIterator, Coroutine
+from dataclasses import dataclass, field
+from datetime import timedelta
+from enum import Enum
+from typing import Self
 
+from aiojobs import Scheduler
+from safir.asyncio import AsyncMultiQueue
+from safir.datetime import current_datetime
+from safir.slack.blockkit import SlackException
 from safir.slack.webhook import SlackWebhookClient
+from sse_starlette import ServerSentEvent
 from structlog.stdlib import BoundLogger
 
 from ..config import LabConfig
-from ..exceptions import MissingSecretError
+from ..constants import LAB_STATE_REFRESH_INTERVAL
+from ..exceptions import (
+    KubernetesError,
+    LabDeletionError,
+    LabExistsError,
+    MissingSecretError,
+    OperationConflictError,
+    UnknownUserError,
+)
 from ..models.domain.docker import DockerReference
 from ..models.domain.gafaelfawr import GafaelfawrUser
+from ..models.domain.kubernetes import PodPhase
+from ..models.domain.lab import LabObjectNames
 from ..models.domain.rspimage import RSPImage
-from ..models.v1.lab import LabSpecification, LabStatus, UserLabState
+from ..models.v1.event import Event, EventType
+from ..models.v1.lab import LabSpecification, LabStatus, PodState, UserLabState
 from ..storage.kubernetes.lab import LabStorage
 from ..storage.metadata import MetadataStorage
 from .builder.lab import LabBuilder
 from .image import ImageService
 from .size import SizeManager
-from .state import LabStateManager
 
 __all__ = ["LabManager"]
 
 
+@dataclass
+class _State:
+    """Collects all internal state for a user's lab."""
+
+    monitor: _LabMonitor
+    """Monitor for any in-progress lab operation."""
+
+    state: UserLabState | None = None
+    """Current state of the lab, in the form returned by status routes."""
+
+    events: AsyncMultiQueue[Event] = field(default_factory=AsyncMultiQueue)
+    """Events from the current or most recent lab operation."""
+
+
+class _LabOperation(Enum):
+    """Possible operations on a lab that could be in progress."""
+
+    SPAWN = "spawn"
+    DELETE = "delete"
+
+
+@dataclass
+class _Operation:
```

Comment:
All these coroutines floating around are making me nervous.

Reply:
It was even worse in earlier versions. I tried about four ways of structuring this before I found something that I could wrap my head around.